File: test/jvm/com/cattailsw/nanidroid/GhostShellNameCompatibilityTest.java
package com.cattailsw.nanidroid;

import static org.junit.Assert.assertEquals;

import java.util.Collections;
import java.util.Map;

import org.junit.Test;
import org.junit.Rule;

/** Guards the legacy distinction between a missing shell descriptor and a missing name. */
public final class GhostShellNameCompatibilityTest {
    @Rule
    public final HostAndroidStubRule androidStubs = new HostAndroidStubRule();
    @Test
    public void missingNameInParsedShellDescriptorRemainsNull() {
        TestGhost ghost = new TestGhost();
        ghost.setShellDescription(Collections.<String, String>emptyMap());

        assertEquals(null, ghost.getShellName());
    }

    @Test
    public void unavailableShellDescriptorFallsBackToMaster() {
        TestGhost ghost = new TestGhost();

        assertEquals("master", ghost.getShellName());
    }

    private static final class TestGhost extends Ghost {
        TestGhost() {
            super("shell-name-contract");
        }

        @Override
        protected void loadGhostInfo() {
            // The test supplies its descriptor state without filesystem or SHIORI setup.
        }

        @Override
        protected void incrementCreateCount() {
            // The test has no persistent creation count.
        }

        void setShellDescription(Map<String, String> description) {
            shellDesc = description;
        }
    }
}

File: src/test/java/com/cattailsw/nanidroid/GhostShellNameCompatibilityTest.kt
package com.cattailsw.nanidroid

import org.junit.Assert.assertEquals
import org.junit.Rule
import org.junit.Test

/** Guards the legacy distinction between a missing shell descriptor and a missing name. */
class GhostShellNameCompatibilityTest {
    @Rule
    @JvmField
    val androidStubs: HostAndroidStubRule = HostAndroidStubRule()

    @Test
    fun missingNameInParsedShellDescriptorRemainsNull() = fixture(shellName = null).use { fixture ->
        assertEquals(null, fixture.requireHandle().ghost.shellName)
    }

    @Test
    fun unavailableShellDescriptorFallsBackToMaster() = fixture(shellName = "master").use { fixture ->
        assertEquals("master", fixture.requireHandle().ghost.shellName)
    }

    private fun fixture(shellName: String?) = RuntimeFixture(
        autoAttach = false,
        preparedFactory = { operationId, ghostId, root ->
            preparedGhost(operationId, ghostId, root, shellName = shellName)
        },
    )
}
